src/versionbits.cpp

```diff
@@ -55,102 +75,85 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex*
         pindexPrev = vToCompute.back();
         vToCompute.pop_back();
 
+        bool finished = pindexPrev->GetMedianTimePast() >= nTimeTimeout;
         switch (state) {
-            case ThresholdState::DEFINED: {
-                if (pindexPrev->GetMedianTimePast() >= nTimeStart) {
+            case ThresholdState::DEFINED:
+                if (finished) {
+                    stateNext = ThresholdState::ABANDONED;
+                } else if (pindexPrev->GetMedianTimePast() >= nTimeStart) {
                     stateNext = ThresholdState::STARTED;
                 }
                 break;
-            }
             case ThresholdState::STARTED: {
-                // We need to count
+                if (finished) {
+                    stateNext = ThresholdState::ABANDONED;
+                    break;
+                }
+                // We need to look for the signal
                 const CBlockIndex* pindexCount = pindexPrev;
-                int count = 0;
-                for (int i = 0; i < nPeriod; i++) {
-                    if (Condition(pindexCount, params)) {
-                        count++;
+                bool sig_active = false;
+                bool sig_abandon = false;
+                for (int i = 0; i < nPeriod; ++i) {
+                    if (pindexCount->nVersion == abandon) {
+                        sig_abandon = true;
+                    } else if (pindexCount->nVersion == activate) {
+                        sig_active = true;
```

Comment:
Isn't this supposed to be counting, or something? Or is it just iterating back to the nPeriod'th block?

Reply:
Changed to `pindexCheck` -- it only cares if it finds a single block in the period that signals for abandonment or (failing that) activation.